internal/fields/mappings.go

```diff
@@ -0,0 +1,577 @@
+		}
+		// logger.Debugf(">>> Comparing field with properties (object): %q", path)
+		compareErrors := compareMappings(path, mappingDefinitions(previewProperties), mappingDefinitions(actualProperties), ecsSchema, localSchema)
+		errs = append(errs, compareErrors...)
+
+		if len(errs) == 0 {
+			return nil
+		}
+		return errs.Unique()
+	}
+
+	if isMultiFields(actual) {
+		if !isMultiFields(preview) {
+			errs = append(errs, fmt.Errorf("not found multi_fields in preview mappings for path: %s", path))
+			return errs.Unique()
+		}
+		previewFields, err := getMappingDefinitionsField("fields", preview)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("found invalid multi_fields type in preview mappings for path %q: %w", path, err))
+		}
+		actualFields, err := getMappingDefinitionsField("fields", actual)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("found invalid multi_fields type in actual mappings for path %q: %w", path, err))
+		}
+		// logger.Debugf(">>> Comparing multi_fields: %q", path)
+		compareErrors := compareMappings(path, mappingDefinitions(previewFields), mappingDefinitions(actualFields), ecsSchema, localSchema)
+		errs = append(errs, compareErrors...)
+		// not returning here to keep validating the other fields of this object if any
+	}
+
+	// Compare all the other fields
+	containsMultifield := isMultiFields(actual)
+	for key, value := range actual {
+		if containsMultifield && key == "fields" {
+			// already checked
+			continue
+		}
+		currentPath := fmt.Sprintf("%s.%s", path, key)
+		if path == "" {
+			currentPath = key
+		}
+		if skipValidationForField(currentPath) {
+			logger.Debugf("Skipped mapping due to path being part of the skipped ones: %s", currentPath)
+			continue
+		}
+
+		// This key does not exist in the preview mapping
+		if _, ok := preview[key]; !ok {
+
+			if childField, ok := value.(map[string]any); ok {
+				if isEmptyObject(mappingDefinitions(childField)) {
+					// TODO: Should this be raised as an error instead?
+					logger.Debugf("field %q is an empty object and it does not exist in the preview", currentPath)
+
+					continue
+				}
+
+				logger.Debugf("Calculating flatten fields for %s", currentPath)
+				flattenFields, err := flattenMappings(currentPath, childField)
+				if err != nil {
+					errs = append(errs, err)
+					return errs
+				}
+
+				for fieldPath, object := range flattenFields {
+					logger.Debugf("- %s", fieldPath)
+
+					def, ok := object.(mappingDefinitions)
+					if !ok {
+						errs = append(errs, fmt.Errorf("invalid field definition/mapping for path: %q", fieldPath))
+						continue
+					}
+
+					if isLocalFieldTypeArray(fieldPath, localSchema) {
+						logger.Debugf("Found field definition with type array, skipping path: %q", fieldPath)
+						continue
+					}
+
+					// TODO: validate mapping with dynamic templates first than validating with ECS
+					// just raise an error if both validation processes fail
```

Comment:
Pending, to be done as part of https://github.com/elastic/elastic-package/issues/2207